docs/development/design/quic-driver.md

```diff
@@ -0,0 +1,156 @@
+- Agent listen on the local interface and encapsulate the received IP packet in the QUIC DATAGRAM and send it to peer agent over QUIC connection
+- Peer agent receives the QUIC DATAGRAM and decapsulate the packet and send it to the local interface
+
+```mermaid
+graph LR
+IP1(IP Packet) <-.-> A(tun0 10.100.0.1) <-.-> NA[Agent\n 10.100.0.1 <-> 10.100.0.2]<-.->I((Internet))
+I((Internet))<-.-> NB[Agent\n 10.100.0.2 <-> 10.100.0.1]
+NB[Agent\n 10.100.0.2 <-> 10.100.0.1]<-.-> B(tun0 10.100.0.1) <-.-> IP2( IP PACKET)
+```
+
+Establishing a QUIC connection across two nodes that's reachable through a public ip address is straight forward. But supporting the following scenarios needs some more work at agent level.
+
+- QUIC connection across two nodes behind reflexive ip address (allowing all inbound/outbound traffic)
+- QUIC connection across two nodes behind easy NAT and Symmetric NAT
+
+QUIC protocol itself is unaware of the type of endpoint address (public ip, reflexive ip, easy NAT), so agents need to ensure that it provides all the details of the peer's ip addresses and port where it can establish the QUIC connection. We will need some of the similar solutions like we use for wireguard (stun base discovery) to support these scenarios.
+
+## Future Considerations
+
+We can implement all the connectivity scenarios that is supported by wireguard through QUIC DATAGRAM tunnel, but similar to wireguard, in case of symmetric NAT it needs to go through relay, or in worst case if gateway firewall disables the UDP protocol for inbound and outbound both, it won't be able to establish the tunnel. Recently approved HTTP DATAGRAM extension ("connect-ip") can be helpful in addressing these scenarios. This extension allows us to transport IP packets over HTTP. HTTP CONNECT will encapsulate packet in HTTP DATAGRAM and forward it to QUIC proxy, if HTTP3/QUIC DATAGRAM is available, it will tunnel traffic using QUIC DATAGRAM, if not, it will fallback to http/2 DATAGRAM. This way, even if UDP is blocked, it can still establish the tunnel through http.
+
+```mermaid
+graph LR
+IP1(IP Packet) <-.-> A(http-proxy/QUIC) <-.->I((Internet))
+I((Internet))<-.-> B(http-proxy/QUIC)
+B(http-proxy/QUIC)<-.-> IP2( IP PACKET)
+```
+
+This solution can also be extended to support the user space proxy use cases as well. Similar to `wireguard-go`, a QUIC aware HTTP proxy can be implemented to support the use space proxy use cases.
+
+### Major challenges in implementing this approach
+
+- HTTP DATAGRAM "connect-ip" extension is not yet supported by any of the QUIC protocol implementations, so that's a major challenge in implementing this approach.
+- Once "connect-ip" support is available, we need http proxies to support this extension, or we can put effort into adding that support to the proxy of our choice.
+  - Existing proxies that support QUIC are listed [here](https://github.com/quicwg/base-drafts/wiki/Implementations)
```

Comment:
Are these the reasons that you didn't propose using `connect-ip` and instead suggest using QUIC DATAGRAM directly?